### ISSUE TITLE
caseOrder Case E must go before T but T must go before E

### ISSUE THREAD

**Issue by sammy-hughes** [go-critic/go-critic#1441]
Go Critic doesn't like generic type-switches. It reports a caseOrder lint on both orderings of a type-switch which asserts on either of two type parameters.

I have a little testing helper meant to enforce a value-or-error-not-both invariant. The constructor takes a type and error, but will only expose one of them, not both. I accomplish that second part with a type-switch asserting on the type of a value, with a value branch and an error/failure branch. The type parameters are called T and E, both with a constraint of `comparable` (I need equality for a zero-value check, and it just makes life easier versus any).

```go
type Maybe[T, E comparable] struct {
	result any
}

func NewMaybe[T, E comparable](ok T, fail E) Maybe[T, E] {
	var zeroNone E
	if fail == zeroNone {
		return Maybe[T, E]{result: ok}
	}

	return Maybe[T, E]{result: fail}
}

func (maybe *Maybe[T, E]) Unwrap() (ok T, fail E) {
	switch result := maybe.result.(type) {
	case T:
		ok = result
	case E:
		fail = result
	}

	return ok, fail
}
```

I have the type-switch ordered as T then E. Go Critic complains. I reorder it as E then T. Go critic complains.

This is clearly disordered.

**Comment by quasilyte** [comment]
The type switches with generic cases should probably be skipped.
Or things like `comparable` should not be treated as interfaces.
I'm not sure how `go/types` reports these.
